CONTRIBUTING.md

```diff
@@ -51,5 +51,104 @@ This process can be performed automatically by running `make license-generator`.
 
+## Developer Guide
+
+The following instructions are intended to help PureScript users more easily contribute to the compiler, even if this is your first Haskell project.
+
+### Prerequisites
+
+Install `stack`. [Instructions](https://docs.haskellstack.org/en/stable/README/).
+
+See the [PureScript Install Guide](INSTALL.md) for information on the `curses`/`ncurses` dependency.
+
+Update stack's package index before proceeding:
+```
+stack update
+```
+
+### Clone
+
+```
+git clone https://github.com/purescript/purescript.git purescript_compiler
+cd purescript_compiler
+```
+
+### Build
+
+```
+stack build
+```
+
+This will take a while the first time it is run.
+
+### Install
+
+If you'd like to test the build in a local shell, you can run:
+```
+stack exec bash
+```
+
+This prepends the location of the compiler to your `$PATH`, so you can test your changes. This works with other tools that depend on the `purs` binary, such as `spago`. Note that rebuilds continue to use this same path, so you don't need to keep re-running `stack exec` to test new changes (assuming you're using a separate terminal for `stack exec`). Remember that `stack exec` must be launched from the same directory that you ran `stack build` from.
+
+You may also install globally with:
+```
+stack install
+```
+
+This copies the compiler executable (`purs`) into `~/.local/bin`.
+
+### Profiling
+
+A profiling build is used to help diagnose performance issues with the compiler.
+
+Create a profiling build with:
+```
+stack build --profile
+```
+This will also take a while the first time it is run.
+
+Setting-up a local shell for your profiling build is similar to the steps for the standard build, just add the `--profile` flag:
+```
+stack exec --profile bash
+```
+Note that the bin directory prepended to `$PATH` is different than the standard build, so you can let this be a third "profiling" shell that you leave open between rebuilds.
+
+The `purs` compiler is often wrapped by `spago`. Here's how to pass the "time profiling" flag `-p` via spago:
+```
+spago build --purs-args "+RTS -p -RTS"
+```
+
+Note: There are other profiling flags (such as `-hc` for heap size). You can read more about these flags [here](http://book.realworldhaskell.org/read/profiling-and-optimization.html).
+
+This creates a `purs.prof` file. You can view the contents of this file directly, but it is often more convenient to use a visualizer.
+
+### Profile Visualizers
+
+Each of these produces a clickable visual display of profiling info. Feel free to open the output files in the web browser of your choice. These examples use `firefox`.
+
+#### [ghc-prof-flamegraph](https://github.com/fpco/ghc-prof-flamegraph)
+```
+stack install ghc-prof-flamegraph
```

Comment:
```diff
- stack install ghc-prof-flamegraph
+ stack build --copy-compiler-tool ghc-prof-flamegraph
```

Reply:
`stack install` copies to:
```
~/.local/bin/
```

`stack build --copy-compiler-tool` copies to:
```
/.stack/compiler-tools/x86_64-linux-tinfo6/ghc-8.10.4/bin/
```

Is there a way to point to the `compiler-tools` location when running the following command in this readme? Is it necessary to have per-compiler versions of this tool?

Reply:
> Is it necessary to have per-compiler versions of this tool?

I'm not sure, but it would remove the surprise that people may have when they find that the tool no longer works when using it with other Haskell projects they have on their local computer. I think the main reason to not use `stack install` is that there isn't a corresponding `stack uninstall`. 

If this change was made, the rest of the command would need to be updated to use `stack exec bash`:
```
stack build --copy-compiler-tool ghc-prof-flamegraph
stack exec bash <<EOF
ghc-prof-flamegraph purs.prof
EOF
firefox purs.svg
```

Reply:
This combo works:
```
stack build --copy-compiler-tool profiteur
stack exec -- profiteur purs.prof
```

Updated instructions.